<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Objectives" data-toc-modified-id="Objectives-1">Objectives</a></span></li><li><span><a href="#This-Document" data-toc-modified-id="This-Document-2">This Document</a></span></li><li><span><a href="#Concepts-&amp;-Tools" data-toc-modified-id="Concepts-&amp;-Tools-3">Concepts &amp; Tools</a></span><ul class="toc-item"><li><span><a href="#Jupyter" data-toc-modified-id="Jupyter-3.1">Jupyter</a></span></li><li><span><a href="#pandas" data-toc-modified-id="pandas-3.2"><em>pandas</em></a></span></li><li><span><a href="#Python-Data-Structures" data-toc-modified-id="Python-Data-Structures-3.3">Python Data Structures</a></span><ul class="toc-item"><li><span><a href="#Lists-&amp;-tuples" data-toc-modified-id="Lists-&amp;-tuples-3.3.1">Lists &amp; tuples</a></span><ul class="toc-item"><li><span><a href="#Basic-definition-and-access" data-toc-modified-id="Basic-definition-and-access-3.3.1.1">Basic definition and access</a></span></li><li><span><a href="#Slicing-to-access-portions" data-toc-modified-id="Slicing-to-access-portions-3.3.1.2">Slicing to access portions</a></span></li><li><span><a href="#Using-for-to-access-a-list:" data-toc-modified-id="Using-for-to-access-a-list:-3.3.1.3">Using <code>for</code> to access a <code>list</code>:</a></span></li><li><span><a href="#Using-for-to-define-a-list:" data-toc-modified-id="Using-for-to-define-a-list:-3.3.1.4">Using <code>for</code> to define a <code>list</code>:</a></span></li></ul></li><li><span><a href="#Python-dictionaries" data-toc-modified-id="Python-dictionaries-3.3.2">Python dictionaries</a></span><ul class="toc-item"><li><span><a href="#Basic-definition-and-access" data-toc-modified-id="Basic-definition-and-access-3.3.2.1">Basic definition and access</a></span></li><li><span><a href="#Using-for-to-access-a-dict" data-toc-modified-id="Using-for-to-access-a-dict-3.3.2.2">Using <code>for</code> to access a <code>dict</code></a></span></li><li><span><a href="#Using-for-to-define-a-dict" data-toc-modified-id="Using-for-to-define-a-dict-3.3.2.3">Using <code>for</code> to define a <code>dict</code></a></span></li></ul></li><li><span><a href="#More-about-lists,-dictionaries,-and-other-data-structures-in-Python" data-toc-modified-id="More-about-lists,-dictionaries,-and-other-data-structures-in-Python-3.3.3">More about lists, dictionaries, and other data structures in Python</a></span></li></ul></li><li><span><a href="#Note-on-Programming-Techniques" data-toc-modified-id="Note-on-Programming-Techniques-3.4">Note on Programming Techniques</a></span></li></ul></li><li><span><a href="#Setup" data-toc-modified-id="Setup-4">Setup</a></span><ul class="toc-item"><li><span><a href="#Displaying-results-in-the-notebook" data-toc-modified-id="Displaying-results-in-the-notebook-4.1">Displaying results in the notebook</a></span></li><li><span><a href="#Reading-Excel-files,-processing-data" data-toc-modified-id="Reading-Excel-files,-processing-data-4.2">Reading Excel files, processing data</a></span></li></ul></li><li><span><a href="#The-Excel-workbook:-2017-CAM-data-from-iPads.xlxs" data-toc-modified-id="The-Excel-workbook:-2017-CAM-data-from-iPads.xlxs-5">The Excel workbook: <code>2017 CAM data from iPads.xlxs</code></a></span><ul class="toc-item"><li><span><a href="#Worksheets-in-the-file" data-toc-modified-id="Worksheets-in-the-file-5.1">Worksheets in the file</a></span></li><li><span><a href="#Unifying-the-DataFrames-(worksheets)" data-toc-modified-id="Unifying-the-DataFrames-(worksheets)-5.2">Unifying the DataFrames (worksheets)</a></span><ul class="toc-item"><li><span><a href="#pandas-functions-for-merging-DataFrames" data-toc-modified-id="pandas-functions-for-merging-DataFrames-5.2.1"><em>pandas</em> functions for merging DataFrames</a></span></li><li><span><a href="#Column-names-differ" data-toc-modified-id="Column-names-differ-5.2.2">Column names differ</a></span><ul class="toc-item"><li><span><a href="#How-shall-we-solve-the-name-mismatch?" data-toc-modified-id="How-shall-we-solve-the-name-mismatch?-5.2.2.1">How shall we solve the name mismatch?</a></span></li><li><span><a href="#How-(where)-should-we-rename-the-columns/indices?" data-toc-modified-id="How-(where)-should-we-rename-the-columns/indices?-5.2.2.2">How (where) should we rename the columns/indices?</a></span></li><li><span><a href="#Renaming-axis-labels-(columns-or-rows)-with-pandas" data-toc-modified-id="Renaming-axis-labels-(columns-or-rows)-with-pandas-5.2.2.3">Renaming axis labels (columns or rows) with <em>pandas</em></a></span></li></ul></li><li><span><a href="#Finally-uniting-our-dataset" data-toc-modified-id="Finally-uniting-our-dataset-5.2.3">Finally uniting our dataset</a></span></li></ul></li><li><span><a href="#Making-sense-of-really-long,-repetitive-column-names" data-toc-modified-id="Making-sense-of-really-long,-repetitive-column-names-5.3">Making sense of really long, repetitive column names</a></span><ul class="toc-item"><li><span><a href="#Nested-dictionaries,-for-a-hierarchy" data-toc-modified-id="Nested-dictionaries,-for-a-hierarchy-5.3.1">Nested dictionaries, for a hierarchy</a></span></li><li><span><a href="#Visualizing-the-tree" data-toc-modified-id="Visualizing-the-tree-5.3.2">Visualizing the tree</a></span></li><li><span><a href="#Accessing-data-using-the-tree" data-toc-modified-id="Accessing-data-using-the-tree-5.3.3">Accessing data using the tree</a></span></li><li><span><a href="#Helper-functions,-for-browsing-the-data" data-toc-modified-id="Helper-functions,-for-browsing-the-data-5.3.4">Helper functions, for browsing the data</a></span><ul class="toc-item"><li><span><a href="#Personal-thought:" data-toc-modified-id="Personal-thought:-5.3.4.1">Personal thought:</a></span></li><li><span><a href="#Improve-the-tree-visualizer-function-with-child_data" data-toc-modified-id="Improve-the-tree-visualizer-function-with-child_data-5.3.4.2">Improve the tree visualizer function with <code>child_data</code></a></span></li></ul></li></ul></li><li><span><a href="#Putting-it-all-together" data-toc-modified-id="Putting-it-all-together-5.4">Putting it all together</a></span><ul class="toc-item"><li><span><a href="#Accessing-data-from-any-section-&amp;-ignoring-irrelevant-rows" data-toc-modified-id="Accessing-data-from-any-section-&amp;-ignoring-irrelevant-rows-5.4.1">Accessing data from any section &amp; ignoring irrelevant rows</a></span></li><li><span><a href="#Outputting-cleaner-data" data-toc-modified-id="Outputting-cleaner-data-5.4.2">Outputting cleaner data</a></span></li></ul></li></ul></li><li><span><a href="#Analyzing-sections-and-subsections" data-toc-modified-id="Analyzing-sections-and-subsections-6">Analyzing sections and subsections</a></span></li><li><span><a href="#Appendix:-Annotated-Data-Structure" data-toc-modified-id="Appendix:-Annotated-Data-Structure-7">Appendix: Annotated Data Structure</a></span><ul class="toc-item"><li><span><a href="#Section:-clients" data-toc-modified-id="Section:-clients-7.1">Section: <code>clients</code></a></span></li><li><span><a href="#Section:-observers" data-toc-modified-id="Section:-observers-7.2">Section: <code>observers</code></a></span></li><li><span><a href="#Section:-fields" data-toc-modified-id="Section:-fields-7.3">Section: <code>fields</code></a></span><ul class="toc-item"><li><span><a href="#fields-Subsection:-client" data-toc-modified-id="fields-Subsection:-client-7.3.1"><code>fields</code> Subsection: <code>client</code></a></span></li><li><span><a href="#fields-Subsection:-oSets" data-toc-modified-id="fields-Subsection:-oSets-7.3.2"><code>fields</code> Subsection: <code>oSets</code></a></span></li><li><span><a href="#oSets-Subsection:-oPoints" data-toc-modified-id="oSets-Subsection:-oPoints-7.3.3"><code>oSets</code> Subsection: <code>oPoints</code></a></span><ul class="toc-item"><li><span><a href="#oPoints-Subsection:-location" data-toc-modified-id="oPoints-Subsection:-location-7.3.3.1"><code>oPoints</code> Subsection: <code>location</code></a></span></li><li><span><a href="#oPoints-Subsection:-observations" data-toc-modified-id="oPoints-Subsection:-observations-7.3.3.2"><code>oPoints</code> Subsection: <code>observations</code></a></span></li></ul></li></ul></li></ul></li></ul></div>

# Objectives

* Read two Excel worksheets
* Merge the data between the two sheets into one
* Demonstrate ability to query the data
* Output at least one Excel file for external analysis that has all relevant information on each row
* Share whatever knowledge is necessary to reproduce these results

# This Document

* You may view or download this notebook on [NBViewer], though it won't be interactive.
* You may interact with this notebook on [mybinder.org], though changes won't be saved for very long.
* The [source] for this notebook is online at GitHub.com, where it can be viewed or downloaded. Instructions on how to use this notebook on your own computer are in the [README].

[![Binder](https://mybinder.org/badge.svg)][mybinder.org]

[README]: https://github.com/devvyn/aafc-field-data/blob/master/README.md
[source]: https://github.com/devvyn/aafc-field-data/blob/master/notebook/Using%20Python%20%26%20pandas%20to%20Analyze%20an%20Excel%20File.ipynb
[mybinder.org]: https://mybinder.org/v2/gh/devvyn/aafc-field-data/master?filepath=notebook%2FUsing%20Python%20%26%20pandas%20to%20Analyze%20an%20Excel%20File.ipynb
[NBViewer]: https://nbviewer.jupyter.org/github/devvyn/aafc-field-data/blob/master/notebook/Using%20Python%20%26%20pandas%20to%20Analyze%20an%20Excel%20File.ipynb

# Concepts & Tools

## Jupyter

## _pandas_

@todo: intro to _pandas_

* Series
* Index
* DataFrame
* I/O

## Python Data Structures

### Lists & tuples

A [`list`] in Python is simply a sequence of objects, which can be numbers, text, or anything else. It uses square brackets.

A [`tuple`] is almost identical to a `list`, except it can't be edited after it's defined. It uses round parentheses. You won't need to worry too much about needing tuples unless you want to make sure the sequence can't be altered.

[`list`]: https://docs.python.org/3/library/stdtypes.html#list
[`tuple`]: https://docs.python.org/3/library/stdtypes.html#tuples

#### Basic definition and access

##### Define a tuple:

In [242]:
my_tuple = (1, 2, 3, "four", ("nested 1", "nested 2"), 30, 0)

##### Define a list:

In [245]:
my_list = [1, 2, 3, "four"]
a_tiny_list = [8]
another_list = [my_list, 5, 6, 7, my_tuple]

In [246]:
my_list

[1, 2, 3, 'four']

In [247]:
a_tiny_list

[8]

In [248]:
another_list

[[1, 2, 3, 'four'],
 5,
 6,
 7,
 (1, 2, 3, 'four', ('nested 1', 'nested 2'), 30, 0)]

##### Access items in a `list` or `tuple`:

In [243]:
my_list[0]

1

In [244]:
my_tuple[0]

1

There's an "item zero". That's normal, because 0 is the origin, just like in Cartesian coordinates.

The index sequence continues as you might expect:

In [172]:
my_list[1]

2

In [173]:
my_list[2]

3

In [174]:
my_list[3]

'four'

In [175]:
my_list[4]

IndexError: list index out of range

If you attempt to access a non existent item by using an index that's out of range, you'll receive an error.

You can, however, use negative numbers! Negative numbers are assumed to be relative to the upper boundary of the range. -1 refers to the _last_ item, -2 to the _second last_, and so on.

In [176]:
my_list[-1]

'four'

In [249]:
my_tuple[-2]

30

##### Adding items to a list:

In [181]:
a_tiny_list

[8]

In [185]:
a_tiny_list.append("another item")

a_tiny_list

[8, 'another item', 'another item']

#### Slicing to access portions

If you want to choose a portion of a list, you can use the colon character to separate the start and end index numbers you want to use as the selected range of the list. The end is where you'll stop, so if there's an item at that index, it'll be excluded:

In [250]:
my_list[0:2]

[1, 2]

In [255]:
my_tuple[5:7]

(30, 0)

You can omit the first or last number in your slice to use the natural upper or lower boundary:

In [179]:
my_list[:2]  # until index 2

[1, 2]

In [256]:
my_tuple[5:]  # index 5 and on

(30, 0)

#### Using `for` to access a `list`:

In [233]:
for item in my_list:
    print(item)

1
2
3
four


In [232]:
for item in another_list:
    print(item)

[1, 2, 3, 'four']
5
6
7


#### Using `for` to define a `list`:

Apply an expression to every item in a list:

In [235]:
my_new_list = [f"I have {item}!" for item in my_list]
my_new_list

['I have 1!', 'I have 2!', 'I have 3!', 'I have four!']

Use `if` to filter a list:

In [237]:
my_filtered_list = [item for item in my_list if item == "four" or item > 2]
my_filtered_list

[3, 'four']

### Python dictionaries

#### Basic definition and access

Dictionaries in Python are just a collection of named things. The things can be another dictionary, a string, a number, or whatver. Even the names of the things don't necessarily have to be words—they can be numbers, for example.

##### Define a dictionary:

In [257]:
my_dictionary = {
    'one': 1,
    2: 'two',
    'green': 'I like colour green.',
    'another dictionary': {'more stuff': 1024,
                           'even more stuff': 2048},
    'a list': [1, 2, 3, 4]
}

In [258]:
my_dictionary

{2: 'two',
 'a list': [1, 2, 3, 4],
 'another dictionary': {'even more stuff': 2048, 'more stuff': 1024},
 'green': 'I like colour green.',
 'one': 1}

##### Access stored values:

In [259]:
my_dictionary["one"]

1

In [260]:
my_dictionary[2]

'two'

In [261]:
my_dictionary["another dictionary"]["even more stuff"]

2048

Note that the attempt to use a non-existant key results in a Python exception called a `KeyError`. This is helpful if you accidentally lose track of which keys are in the dictionary. 

In [262]:
my_dictionary["some key that doesn't exist"]

KeyError: "some key that doesn't exist"

If you don't want this to happen, there are ways to get around it. Later on, we'll use something called a [`defaultdict`], which just makes up the new key on the spot, instead of stopping dead.

[`defaultdict`]: https://docs.python.org/3/library/collections.html?highlight=defaultdict#defaultdict-objects

##### Assign values to existing keys or new keys:

In [263]:
my_dictionary["two"] = my_dictionary[2]
my_dictionary[2] = "fourteen"

In [264]:
my_dictionary["some key that didn't exist before"] = "Some value that didn't exist before"

In [265]:
my_dictionary.keys()

dict_keys(['one', 2, 'green', 'another dictionary', 'a list', 'two', "some key that didn't exist before"])

#### Using `for` to access a `dict`

##### Dictionary iterables

Python's `for` statements are handy for doing something to each item in a collection. Dictionaries are a type of collection, but since they have keys and values, you need to specify which you want. To address this, there are three handy methods that all dictionaries have:

- [`keys()`] - Only the key names
- [`values()`] - Only the values
- [`items()`] - Pairs of keys and values

[`keys()`]: https://docs.python.org/3/library/stdtypes.html#dict.keys
[`values()`]: https://docs.python.org/3/library/stdtypes.html#dict.values
[`items()`]: https://docs.python.org/3/library/stdtypes.html#dict.items



##### `keys()`

In [266]:
for key in my_dictionary.keys():
    display(key)

'one'

2

'green'

'another dictionary'

'a list'

'two'

"some key that didn't exist before"

##### `values()`

In [267]:
for value in my_dictionary.values():
    display(value)

1

'fourteen'

'I like colour green.'

{'even more stuff': 2048, 'more stuff': 1024}

[1, 2, 3, 4]

'two'

"Some value that didn't exist before"

##### `items()`

In [268]:
for item in my_dictionary.items():
    display(item)

('one', 1)

(2, 'fourteen')

('green', 'I like colour green.')

('another dictionary', {'even more stuff': 2048, 'more stuff': 1024})

('a list', [1, 2, 3, 4])

('two', 'two')

("some key that didn't exist before", "Some value that didn't exist before")

#### Using `for` to define a `dict`

With a list of items, it's often useful to make a dictionary with keys or values based those items. For example, let's say we have this list:

In [221]:
my_list[:3]

[1, 2, 3]

Simply use `for` to access the list and make the dictionary at the same time. You can use expressions to transform the values as you apply them to the new dictionary. The syntax is almost the same as for regular dictionary definition:

In [269]:
my_new_dict = {f"Key #{item}": item + 100
               for item in my_list
               if item != "four"}

In [270]:
my_new_dict

{'Key #1': 101, 'Key #2': 102, 'Key #3': 103}

This is the same as using the regular assignment syntax in a regular `for` loop:

In [271]:
my_new_dict = {}
for item in my_list:
    if item != "four":
        my_new_dict[f"Key #{item}"] = item + 100

In [272]:
my_new_dict

{'Key #1': 101, 'Key #2': 102, 'Key #3': 103}

### More about lists, dictionaries, and other data structures in Python

If you want to learn more about using dictionaries, see the official [Python tutorial on dictionaries], which is part of the page demonstrating other Python collections, such as [lists], which we'll be using extensively.

[Python tutorial on dictionaries]: https://docs.python.org/3/tutorial/datastructures.html#dictionaries
[lists]: https://docs.python.org/3/tutorial/datastructures.html#more-on-lists

## Note on Programming Techniques

@todo: OOP vs FP vs procedural

# Setup

To work with the source data, we need to be able to:

* open an Excel file
* display results of intermediate processing in the notebook
* save results to an Excel file

Most functions to accomplish our goals have been published already. We only need to import them.

## Displaying results in the notebook

You can use the regular [`print()`] function in Python, but you'll always get plain text output. [`display()`] can do plain text as well as HTML, images, audio, etc.

[`print()`]: https://docs.python.org/3/library/functions.html#print 
[`display()`]: http://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html

You may need all the core display objects imported before using [`display()`]. I found I didn't need to import them on new versions of Jupyter Notebook, but since the display function for isn't always imported by default in some implementations, you can run:

In [10]:
from IPython.core.display import *

Here are simple helpers to make headings and other embellishments easy. There are probably libraries already to make HTML tags, but this should suffice, without importing a large library. In general, these methods are unnecessary, but I'll be using them later on in this document to make code output more understandable.

In [3]:
def html_wrap(content, element="span", attributes_str=None):
    """ Convenience method for wrapping any string in any HTML tag. """
    if attributes_str:
        element_open = f"{element} {attributes_str}"
    else:
        element_open = element
    return f"<{element_open}>{content}</{element}>"


def make_heading(content, level=5):
    """ Convenience method for wrapping any string in an HTML header tag. """
    element = f"h{level}"
    return html_wrap(content, element)


def heading(*args, **kwargs):
    """ Call IPython.core.display.HTML on the results of the make_heading
    function, so users don't have to repeatedly do `HTML(heading(…))` """
    return HTML(make_heading(*args, **kwargs))

## Reading Excel files, processing data

[_pandas_] is really good with columnar data, such as the kind of spreadsheets Excel works with.

[_pandas_]: https://pandas.pydata.org

In [2]:
import pandas

_pandas_ also has the ability to read and write Excel files, as long as the `xlrd` distribution is installed. This notebook already has _pandas_ and `xlrd`, but if you need to install them, you can run this command in a Jupyter Notebook cell, like so:

In [13]:
# Install xlrd to the Python environment, if it's not already installed.
!pip install pandas xlrd

Depending on how your Jupyter host is deployed, you may be able to add `xlrd` and `pandas` to `requirements.txt` or `environment.yml`, as appropriate for your configuration.

If you prefer, you can install this from the regular command line shell on your Jupyter host by running the `pip` command without the exclamation mark.

# The Excel workbook: `2017 CAM data from iPads.xlxs`

In [5]:
# The file I'm interested in parsing for cleanup:
file_path = "./data/real/2017/CAM data from iPads/2017 CAM data from iPads.xlsx"

## Worksheets in the file

In [8]:
data_file = pandas.ExcelFile(file_path)
data_file_sheet_names = sorted(data_file.sheet_names)
data_file_sheet_names

['2017 CAM data Erl',
 '2017 CAM iPad data Tyler',
 'Combined iPad 2017 CAM data',
 'schema (WIP reverse engineer)']

I'm only interested in first two, which should have all the data. The other sheets are either unfinished or unrelated to the task at hand.

In [7]:
cam_sheet_names = data_file_sheet_names[:2]
cam_sheet_names

['2017 CAM data Erl', '2017 CAM iPad data Tyler']

Let's make a dictionary of DataFrames from all the sheets, using the last word as the name. If you're familiar with [lists] in Python, you may remember that an index of -1 means the last item, as in the largest possible index number.

[lists]: https://docs.python.org/3/tutorial/datastructures.html#lists

In [9]:
def last_word(string, word_separator=' '):
    words = string.split(word_separator)
    return words[-1]

In [10]:
sheets = {last_word(sheet_name):
          data_file.parse(sheet_name)
          for sheet_name in cam_sheet_names}

In [15]:
sheets.keys()

dict_keys(['Erl', 'Tyler'])

## Unifying the DataFrames (worksheets)

Now that we have a convenient list of sheets that are loaded as _pandas_ DataFrames, we can work toward merging them into one. Once they're merged, we can process the data more easily from a single DataFrame.

### _pandas_ functions for merging DataFrames

_pandas_ has multiple methods for combining datasets:

- [`concat`]: The concatenate method has an option to ignore row numbers, which has the effect of gluing each dataframe to the bottom of the previous one.
- [`append`]: Append would be almost equivalent, but always updates the first DataFrame or Series it's given.
- [`merge`]: Merge is more for database-style relational merges.

[`concat`]: https://pandas.pydata.org/pandas-docs/stable/merging.html
[`append`]: https://pandas.pydata.org/pandas-docs/stable/merging.html#concatenating-using-append
[`merge`]: https://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging

From reading up on all three methods, it looks to me like `concat` will suffice, here, as long as the column names are identical. I like to have the option of creating a new DataFrame rather than overwriting anything, because the results are easier to repeat.

Let's see how close that concatenation method gets us:

In [20]:
pandas.concat(sheets).info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 4690 entries, (Erl, 0) to (Tyler, 3779)
Data columns (total 56 columns):
clients__company                                                    1 non-null object
clients__displayText                                                1 non-null object
clients__fname                                                      1 non-null object
clients__lname                                                      1 non-null object
clients__name                                                       1 non-null object
fields__client__company                                             21 non-null object
fields__client__displayText                                         21 non-null object
fields__client__fname                                               21 non-null object
fields__client__lname                                               21 non-null object
fields__client__name                                                21 non-null object
fields__crop

We can see the RangeIndex size in the info: 

>MultiIndex: 4690 entries, (Erl, 0) to (Tyler, 3779)

Which means 4690 "rows", if this were a spreadsheet. Compare to the total row count for all sheets, by using Python's function `len` (an abbreviation for "length"), which counts the number of items (in this case, rows):

In [21]:
sum(len(sheet) for sheet in sheets.values())

4690

Also in the `info()` readout:

>Data columns (total 56 columns):

Compare to the column count for the source worksheets:

In [22]:
[len(sheet.columns) for sheet in sheets.values()]

[50, 50]

So, 50 columns in each source sheet.

Pretty good, so far, except for some extra columns due to variations in the column names—I see `growthStage` and `growthStage Zadoks`, as well as some aphid names amongst the `a1__number`, `a2__number`, `a3__number` group:

```
fields__oSets__growthStage                                  45 non-null float64
fields__oSets__growthStage Zadoks                           9 non-null float64
```

```
fields__oSets__oPoints__observations__a1__number            324 non-null float64
fields__oSets__oPoints__observations__a1__number EGA        124 non-null float64
fields__oSets__oPoints__observations__a2__number            112 non-null float64
fields__oSets__oPoints__observations__a2__number BCO        30 non-null float64
fields__oSets__oPoints__observations__a3__number            25 non-null float64
fields__oSets__oPoints__observations__a3__number Greenbug   0 non-null float64
```

The total number of rows is correct, and the other columns line up when the names match.

Column names (which become Series names in _pandas_) are the only problem standing in the way of a successful unification of DataFrames.

### Column names differ

It's already pretty obvious what the naming problem is. But just to demonstrate how to use Python to calculate differences, let's look again at our DataFrames column names.

For our frames:

In [23]:
sheets.keys()

dict_keys(['Erl', 'Tyler'])

Calculate the difference in column names, using "Erl" as the base, and Python's set theory models:

In [24]:
display(set.difference(*[set(sheet.columns) for sheet in sheets.values()]))

{'fields__oSets__growthStage Zadoks',
 'fields__oSets__oPoints__observations__a1__number EGA',
 'fields__oSets__oPoints__observations__a2__number BCO',
 'fields__oSets__oPoints__observations__a3__number Greenbug',
 'fields__oSets__oPoints__observations__anum TotalAPhids',
 'fields__oSets__oPoints__observations__eVnum Natural enemy totals'}

Or, if we want to list the names of columns from *all* sheets that don't have a match (symmetric difference):

In [25]:
column_list = sorted(set.symmetric_difference(*[set(sheet.columns) for sheet in sheets.values()]))
display(column_list)

['fields__oSets__growthStage',
 'fields__oSets__growthStage Zadoks',
 'fields__oSets__oPoints__observations__a1__number',
 'fields__oSets__oPoints__observations__a1__number EGA',
 'fields__oSets__oPoints__observations__a2__number',
 'fields__oSets__oPoints__observations__a2__number BCO',
 'fields__oSets__oPoints__observations__a3__number',
 'fields__oSets__oPoints__observations__a3__number Greenbug',
 'fields__oSets__oPoints__observations__anum',
 'fields__oSets__oPoints__observations__anum TotalAPhids',
 'fields__oSets__oPoints__observations__eVnum',
 'fields__oSets__oPoints__observations__eVnum Natural enemy totals']

Let's pop it into a two column layout for _pandas_ to display. I found a handy list-to-grid [recipe] on the official documentation for Python's `itertools` library. I based my function on their `grouper` example.

[recipe]: https://docs.python.org/3/library/itertools.html#itertools-recipes

In [26]:
def grid(items, width=2):
    """ Layout items in a grid, from left to right, top to bottom. """
    return [*zip(*[iter(items)] * width)]

In [27]:
column_grid = pandas.DataFrame(grid(column_list))

In [28]:
# Display entire column, even if cell data is long
pandas.set_option('display.max_colwidth', 0)  # Zero means no limit

# Show the differing column names side-by-side
display(column_grid)

0  \
0  fields__oSets__growthStage                         
1  fields__oSets__oPoints__observations__a1__number   
2  fields__oSets__oPoints__observations__a2__number   
3  fields__oSets__oPoints__observations__a3__number   
4  fields__oSets__oPoints__observations__anum         
5  fields__oSets__oPoints__observations__eVnum        

                                                                  1  
0  fields__oSets__growthStage Zadoks                                 
1  fields__oSets__oPoints__observations__a1__number EGA              
2  fields__oSets__oPoints__observations__a2__number BCO              
3  fields__oSets__oPoints__observations__a3__number Greenbug         
4  fields__oSets__oPoints__observations__anum TotalAPhids            
5  fields__oSets__oPoints__observations__eVnum Natural enemy totals

Transpose the frame by viewing the `T` attribute of our DataFrame, so differences are vertically adjacent:

In [29]:
display(column_grid.T)

0  \
0  fields__oSets__growthStage          
1  fields__oSets__growthStage Zadoks   

                                                      1  \
0  fields__oSets__oPoints__observations__a1__number       
1  fields__oSets__oPoints__observations__a1__number EGA   

                                                      2  \
0  fields__oSets__oPoints__observations__a2__number       
1  fields__oSets__oPoints__observations__a2__number BCO   

                                                           3  \
0  fields__oSets__oPoints__observations__a3__number            
1  fields__oSets__oPoints__observations__a3__number Greenbug   

                                                        4  \
0  fields__oSets__oPoints__observations__anum               
1  fields__oSets__oPoints__observations__anum TotalAPhids   

                                                                  5  
0  fields__oSets__oPoints__observations__eVnum                       
1  fields__oSets__oPoints__observations__eVnum Natural enemy totals

In [30]:
# Reset the column width, since we're done looking at the names this way
pandas.reset_option('display.max_colwidth')

Before we concatenate, we'll standardize on the names from the first sheet (0), since it's more regular. But which sheet is the one to fix? Which sheet has `fields__oSets__growthStage Zadoks` instead of `fields__oSets__growthStage`? I believe "Erl" was our base for comparison initially, but let's make sure of what we're about to do:

In [31]:
# make a filtered dictionary, for later reference
sheets_with_bad_column_names = {sheet_name: sheet for sheet_name, sheet in sheets.items() 
                                if 'fields__oSets__growthStage Zadoks' in sheet.columns}
display(heading('Bad:'),
        set(sheets_with_bad_column_names.keys()))

{'Erl'}

In [32]:
display(heading('Good:'),
        set(sheets.keys()) - set(sheets_with_bad_column_names.keys()))

{'Tyler'}

Okay, "Tyler" has the column names we prefer, "Erl" does not. Duly noted.

There's now a `sheets_with_bad_column_names` variable we can use when we fix that. (It's only got one DataFrame, but it's still good practice to be ready for batch processing in case this code is reused in the future.)

#### How shall we solve the name mismatch?

Our options:

- rename the columns
- try to merge the sheets in a way that ignores the column names of the bad sheet

If we try to do a special merge, ignoring the column names, we have to worry about the _precise order_ of those columns instead of relying on the names. Since we'd have to spend time checking—and possibly rearranging—the column sequence, we might as well spend that time creating a _reusable_, documented solution for quick and painless name fixing.

#### How (where) should we rename the columns/indices?

Possibilities:

- fix the spreadsheet document in Excel format, and move ahead as if this had never happened
- keep the names in the file, but correct the names in memory once they become indices

Since this notebook you're reading has already mentioned the problem, let's go ahead and solve it here. That way our notebook will detail the fix and how to use it in the future, if there are more Excel spreadsheets with similarly altered names.

#### Renaming axis labels (columns or rows) with _pandas_

_pandas_ has a [`rename`] method which lets us apply a transform function to the indices named after our worksheet columns (or explicitly map each column through a dictionary).

[`rename`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html

Since the bad names merely have extra words tacked onto the end, let's just split the name and use the first "word". In Python, we get the first item of a sequence by using index zero:

In [33]:
def first_word(string, word_separator=' '):
    """ Split string into words (by space character), return first word. """
    return string.split(word_separator)[0]

We just need to apply this function to every Series in the DataFrame, through the _pandas_ rename function. For the sake of visualizing the changes, let's also report on changes in a dry run before using the _pandas_ rename function.

In [34]:
report = []  # For the lines of the report which we'll display later.

for sheet_name, sheet in sheets_with_bad_column_names.items():
    report.append(make_heading(sheet_name, level=3))  # heading
    report_lines = []  # the body of the report for each DataFrame
    for column_name in sheet.columns:
        after = first_word(column_name)
        if after == column_name:
            content = column_name
            attributes = None
        else:
            content = f"{column_name} &rarr; {after}"  # old --> new
            attributes = "style='font-weight: bold'"
        report_lines.append(html_wrap(content, 'li', attributes))  # add an HTML list item
    report.append(html_wrap(''.join(report_lines), 'ol')) # add an ordered list of items to the report
    
# join all the strings together and display as HTML
display(HTML(''.join(report)))

Actually change the names:

In [35]:
for sheet in sheets_with_bad_column_names.values():
    sheet.rename(mapper=first_word, axis='columns', inplace=True)

As mentioned before, we know there's only one sheet, so the `for` loop isn't absolutely necessary, but it's still a good habit when dealing with reusable scripts that can work on batches of datasets. (Sometimes other people find your work useful, and sometimes you'll revisit your work to copy something that succeeded in the past.)

### Finally uniting our dataset

In [36]:
data = pandas.concat(sheets)

In [37]:
data.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 4690 entries, (Erl, 0) to (Tyler, 3779)
Columns: 50 entries, clients__company to observers
dtypes: datetime64[ns](1), float64(27), object(22)
memory usage: 1.8+ MB


Splendid! That's what we expected. 😁

## Making sense of really long, repetitive column names

_pandas_ has a concept of [indexing hierarchically], which may help group columns together. However, I'm new to _pandas_ and I'm not new to Python. Furthermore, from what I've read, if we use a multi-level labelling system for grouping, we have to pay attention to which level we're operating on as we work with the spreadsheet as DataFrame. 

To avoid complications due to my own ignorance, let's use a nested dictionary in Python to accomplish the same thing without _pandas_. We can probably use the dictionary to help us add more indices to the DataFrame later, if we need.

[indexing hierarchically]: https://pandas.pydata.org/pandas-docs/stable/advanced.html

### Nested dictionaries, for a hierarchy

Instead of a regular dictionary in Python, let's use something called a [`defaultdict`], which will simplify the creation of our hierarchy. [`defaultdict`] is like a regular dictionary, except it doesn't complain if you try to access a key that doesn't exist yet—it just adds it.

[`defaultdict`]: https://docs.python.org/3/library/collections.html#collections.defaultdict

Because there's always the chance that a node might have data as well as more nodes under it, we'll store the reference to data in a key that can't possibly be a segment in the name: the separator (which in this case is '__') 

Before we forget, we should save that separator for quick reference:

In [38]:
SEPARATOR = '__'
display(SEPARATOR)

'__'

I don't know about you, but I'll feel weird referring to the data Series as `SEPARATOR` all the time. Let's make an alias, so it's easier to remember. Just because we're storing data under a label that _happens_ to also be the separator for the split we did earlier, doesn't mean we humans have to keep track of that fact long after it's no longer important.

In [39]:
DATA = SEPARATOR

The following method looks long but it's mostly comments for your benefit. 

In [40]:
from collections import defaultdict

def split_column_to_dict(column, column_name=None, column_dictionary=None, separator=SEPARATOR):
    """ Split the column names like "fields__oSets__oPoints__observations" into groupings of keys
    so that related keys are easy to find, ie columns['fields']['oSets']['oPoints']['observations'].
    This produces a tree of column name segments, with references to actual data at the ends.
    
    If `column_name` is provided, it's used instead of the actual column name.
    
    If `column_dictionary` is provided, attempt to add to it as if it were already initialized as
    a nested defaultdict, from an earlier call to this function. """

    # If a dictionary is not provided, make an empty one.
    if column_dictionary is None:
        def nested_dict():
            """ This function will be called by defaultdict
            whenever a non-existent key is used. """
            return defaultdict(nested_dict)
        
        column_dictionary = nested_dict()

    # Set a pointer to the root of the tree, for starters.
    pointer = column_dictionary
    
    # Now, walk through the segments in order from left to right,
    # touching the tree for each one.
    for segment in str(column_name or column.name).split(separator):
        pointer = pointer[segment]
        
    # Now that the loop is done, the pointer is pointing at the deepest
    # level of the branch, which either already existed or else it was created.
    
    # At the end of the branch, put the data under a special key.
    pointer[separator] = column

    # Since `pointer` was actually just pointing to parts of `column_dictionary`,
    # the column dictionary has been filled out with nodes because of how defaultdict
    # was setup with our `nested_dict` constructor.
    return column_dictionary


Just to demonstrate that the functional code is actually quite slim and simple, here it is without any comments:

```python
def split_column_to_dict(dataframe, column_name=None, column_dictionary=None, separator=SEPARATOR):
    if column_dictionary is None:
        def nested_dict():
            return defaultdict(nested_dict)
        column_dictionary = nested_dict()
    pointer = column_dictionary
    for segment in str(column_name or column.name).split(separator):
        pointer = pointer[segment]
    pointer[separator] = dataframe[column_name]
    return column_dictionary
```

Now, make a dictionary of column trees:

In [41]:
column_dictionary = None  # initialize the variable, otherwise we can't reference it
for name, column in data.items():
    column_dictionary = split_column_to_dict(column, name, column_dictionary)

Optionally, we can also insist that from now on, our dictionary should not blithely create keys:

In [42]:
# Turn off the defaultdict behaviour, so errors are easier to detect later on
column_dictionary.default_factory = None

If that worked as planned, there should be a list of the first segments of all the column names in that sheet:

In [43]:
column_dictionary.keys()

dict_keys(['clients', 'fields', 'observers'])

Continuing deeper, more segments that share a common prefix:

In [44]:
fields_node = column_dictionary['fields']
fields_node.keys()

dict_keys(['client', 'crop', 'date', 'desc', 'image', 'name', 'oSets'])

In [45]:
sets_node = fields_node['oSets']
sets_node.keys()

dict_keys(['completeSets', 'date', 'dateCompare', 'desc', 'growthStage', 'oPoints', 'obsName', 'results', 'totalA1', 'totalA2', 'totalA3', 'totalA4', 'totalSets'])

### Visualizing the tree

Browsing the column structure by using the dictionary accessors is a bit tedious. It's great for machines, but my human brain has a bit of trouble holding it all in my head when I see it only one level at a time.

We can see the structure of the hierarchy in a dictionary by using a function like this:

In [46]:
def visualize_tree(tree, depth=1, maximum_depth=None, data_key=DATA):
    if isinstance(tree, (dict, defaultdict)):
        for key, val in tree.items():
            if key != data_key:
                print(".   " * (depth - 1) + key)
                if maximum_depth is None or depth <= maximum_depth:
                    visualize_tree(val, depth + 1, maximum_depth, data_key,
                                   show_leaf_nodes)

It may seem strange to see math symbols used with text, but Python understands what we want: to repeat the text if we multiply it, to append if we add.

Here's the output for the whole tree:

In [47]:
visualize_tree(column_dictionary)

clients


NameError: name 'show_leaf_nodes' is not defined

Using this new function, we can also limit the depth, even to zero, which is the equivalent of `column_dictionary.keys()`:

In [48]:
column_dictionary.keys()

dict_keys(['clients', 'fields', 'observers'])

In [49]:
visualize_tree(column_dictionary, maximum_depth=0)

clients
fields
observers


Or, just a shallow peek into the next level:

In [50]:
visualize_tree(column_dictionary, maximum_depth=1)

clients


NameError: name 'show_leaf_nodes' is not defined

### Accessing data using the tree

At the end of each branch in the tree should be a '\_\_' key for the actual data. For example, the `date` of the set:

In [51]:
sets_node['date'].keys()

dict_keys(['__'])

And we can check for that any time by using Python's `in` operator:

In [52]:
DATA in sets_node['date']

True

We can look at some data now!

In [53]:
display(column_dictionary['observers'][DATA].count())

3

In [54]:
sets_node['date'][DATA].head(20)

Erl  0     2017-08-02T13:12:09.542
     1                         NaN
     2                         NaN
     3                         NaN
     4                         NaN
     5                         NaN
     6                         NaN
     7                         NaN
     8                         NaN
     9                         NaN
     10                        NaN
     11                        NaN
     12                        NaN
     13                        NaN
     14                        NaN
     15                        NaN
     16                        NaN
     17                        NaN
     18                        NaN
     19                        NaN
Name: fields__oSets__date, dtype: object

So, this is the problem with which we launched this whole endeavour. Many, many empty cells. If we want to skip over blanks as we scan down a column, _pandas_ has a [`dropna()`] method for that. Let's use `head` to peek at the first bit of that:

[`dropna()`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dropna.html

In [55]:
sets_node['date'][DATA].dropna().head(10)

Erl    0      2017-08-02T13:12:09.542
       70     2017-08-09T09:25:11.710
       140    2017-08-09T10:06:25.480
       210    2017-08-09T11:21:01.555
       350    2017-08-09T11:37:20.862
       490    2017-08-22T15:42:05.751
       560    2017-08-17T11:12:02.820
       700    2017-08-17T13:06:30.183
       840    2017-08-22T16:02:50.682
Tyler  0      2017-07-14T12:31:24.194
Name: fields__oSets__date, dtype: object

So much better!

We now have enough information to find the beginning of chunks in the sheet by scanning for non-blank cells: we have those row numbers for each of the non-null records of that column, where related columns will also reveal data we need for that section.

As a side note, the number of non-null (not empty) records in any given column was displayed when we called [`info()`] on the DataFrame. We can also get information like that from the column (which is a Series in _pandas_) by using [`describe()`] or [`count()`]:

[`info()`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.info.html
[`describe()`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.describe.html
[`count()`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.count.html

In [56]:
sets_node['date'][DATA].describe()

count                          54
unique                         54
top       2017-08-22T16:02:50.682
freq                            1
Name: fields__oSets__date, dtype: object

### Helper functions, for browsing the data

Once we start actually reading groups of columns based on which type of chunk we're reading from, we'll have to start typing lists of column names for each section.

It's a bit of extra work to pay attention to whether a given node in the hierarchy is holding a reference to a data Series (column) or whether it's only an intermediate step on the way to the end of the branch.

To save time and mental energy, we can filter columns by whether they have data or not, if we make some simple filter functions. For detecting data that we've placed in the hierarchy of columns, we can look for the special key that we chose earlier: the separator, which is two underscore characters (`__`).

In [57]:
def has_data(node, data_key=DATA):
    """ We know node has data if it has a key that's the data key. """
    return isinstance(node, dict) and data_key in node.keys()

def child_data(node):
    """ Filter children that have data. """
    return {parent_key: child for parent_key, child in node.items()
            if has_data(child)}

def child_children(node):
    """ Filter children that have children. We know an child has children if
    it has any values that are dictionaries with at least one key that isn't
    a data reference. """
    return {parent_key: child for parent_key, child in node.items()
            if isinstance(child, dict) and len(child) > 1 or len(child) == 1 and not has_data(child)}

In [58]:
points_node = sets_node['oPoints']
observations_node = points_node['observations']

In [59]:
display(heading('children with children:'),
        child_children(observations_node).keys())

dict_keys(['a1', 'a2', 'a3', '|'])

In [60]:
display(heading('children with data:'),
        child_data(observations_node).keys())

dict_keys(['anum', 'complete', 'disabled', 'eVnum', 'enum', 'id', 'name', '|'])

In [61]:
# With Python sets, the ampersand (`&`) is the set intersetion operator.
display(heading('children with children & data:'),
        set(child_children(observations_node).keys()) & set(child_data(observations_node).keys()))

{'|'}

Oh? Children and data under `|`? What could it be?

In [62]:
display(observations_node['|'][DATA].dropna().head(18))

Erl  5     e1
     6     e2
     7     e3
     8     e4
     9     e5
     10    e6
     11    e7
     12    e8
     13    e9
     19    e1
     20    e2
     21    e3
     22    e4
     23    e5
     24    e6
     25    e7
     26    e8
     27    e9
Name: fields__oSets__oPoints__observations__|, dtype: object

In [63]:
display(observations_node['|']['number'][DATA].dropna().head(18))

Erl  90     5.0
     94     1.0
     96     2.0
     104    5.0
     110    1.0
     118    4.0
     132    2.0
     136    1.0
     145    1.0
     152    1.0
     160    1.0
     166    1.0
     502    2.0
     509    1.0
     544    1.0
     558    2.0
     572    7.0
     586    6.0
Name: fields__oSets__oPoints__observations__|__number, dtype: float64

Based on my experience using the app, and analyzing other datasets, I think it has something to do with natural enemy counts. As far as I've been able to discern, the source app's data doesn't distinguish between a count of zero and unreported (missing) data points. That's why the `observations__|__number` data (which doesn't necessarily exist) isn't always present on the same rows as all the `observations__|` labels.

Later on, we'll need to dig up the actual list of naturaly enemies and match them to the corresponding "e" labels, `e1` through `e9`.

#### Personal thought:

I have to wonder why the developers of the app that output this data chose to use an unpronounceable column name, and put something important there. Not only data, but also another, related column with more data. Why not `enemy__name` and `enemy__number`?

Regardless of how questionable that is, or how that happened, we can handle it now.

#### Improve the tree visualizer function with `child_data`

Now that we have filters for helping us traverse the tree, let's take a moment to improve `visualize_tree`:

In [64]:
def visualize_tree(tree, depth=1, maximum_depth=None, data_key=DATA,
                   show_leaf_nodes=True):
    if isinstance(tree, (dict, defaultdict)):
        if show_leaf_nodes:
            nodes = tree
        else:
            nodes = child_children(tree)
        for key, val in nodes.items():
            if key != data_key:
                if has_data(val):
                    prefix = '* '
                else:
                    prefix = ''
                if child_children(val) or child_data(val):
                    suffix = '/'
                else:
                    suffix = ''
                print(".   " * (depth - 1) + f"{prefix}{key}{suffix}")
                if maximum_depth is None or depth <= maximum_depth:
                    visualize_tree(val, depth + 1, maximum_depth, data_key,
                                   show_leaf_nodes)

Now we can visualize the structure in terms of what's at the ends of the branches (leaf nodes) or all descendants. The additional punctuation also makes it more obvious which nodes have data values. For example, the node at `fields__oSets__oPoints__observations__|`, which is both a data node and a branch.

In [65]:
visualize_tree(observations_node, show_leaf_nodes=False)

a1/
a2/
a3/
* |/


In [66]:
visualize_tree(observations_node, show_leaf_nodes=True)

a1/
.   * number
a2/
.   * number
a3/
.   * number
* anum
* complete
* disabled
* eVnum
* enum
* id
* name
* |/
.   * number


## Putting it all together

### Accessing data from any section & ignoring irrelevant rows

Let's take a look at the columns containing data about observation sets.

We'll use `has_data` to filter our `sets_node` so we can express a list of names of data-containing columns about `fields__oSets`.

In [67]:
sets_columns_names = [column[DATA].name for column in child_data(sets_node).values()]
display(sets_columns_names)

['fields__oSets__completeSets',
 'fields__oSets__date',
 'fields__oSets__dateCompare',
 'fields__oSets__desc',
 'fields__oSets__growthStage',
 'fields__oSets__obsName',
 'fields__oSets__results',
 'fields__oSets__totalA1',
 'fields__oSets__totalA2',
 'fields__oSets__totalA3',
 'fields__oSets__totalA4',
 'fields__oSets__totalSets']

Passing that list to _pandas_, we should get exactly which columns of data we need:

In [68]:
display(data[sets_columns_names].head())

fields__oSets__completeSets      fields__oSets__date  \
Erl 0                          0.0  2017-08-02T13:12:09.542   
    1                          NaN                      NaN   
    2                          NaN                      NaN   
    3                          NaN                      NaN   
    4                          NaN                      NaN   

      fields__oSets__dateCompare  fields__oSets__desc  \
Erl 0                 2017-08-02                  NaN   
    1                        NaT                  NaN   
    2                        NaT                  NaN   
    3                        NaT                  NaN   
    4                        NaT                  NaN   

       fields__oSets__growthStage fields__oSets__obsName  \
Erl 0                         7.0                  Tyler   
    1                         NaN                    NaN   
    2                         NaN                    NaN   
    3                         NaN                    NaN   
    4                         NaN                    NaN   

      fields__oSets__results  fields__oSets__totalA1  fields__oSets__totalA2  \
Erl 0                    NaN                     NaN                     NaN   
    1                    NaN                     NaN                     NaN   
    2                    NaN                     NaN                     NaN   
    3                    NaN                     NaN                     NaN   
    4                    NaN                     NaN                     NaN   

       fields__oSets__totalA3  fields__oSets__totalA4  \
Erl 0                     NaN                     NaN   
    1                     NaN                     NaN   
    2                     NaN                     NaN   
    3                     NaN                     NaN   
    4                     NaN                     NaN   

       fields__oSets__totalSets  
Erl 0                       1.0  
    1                       NaN  
    2                       NaN  
    3                       NaN  
    4                       NaN

Once again, the problem we faced at the outset. How about skipping irrelevant rows? We fixed this earlier with `dropna()`, but this time we're operating on some columns that _may_ be null, plus certain ones that _must not_ be null.

Let's use the date column as the crucial record upon which we'll predicate our filter, because it should never be null. 

_pandas_ has a way to filter on conditions, which is sometimes called [boolean indexing] because the condition is either `True` or `False`. In this case, _pandas_ expects the actual column data object itself (Series), rather than just the name of the column to check. We already have that in the hierarchy from earlier:

[boolean indexing]: https://pandas.pydata.org/pandas-docs/stable/10min.html#boolean-indexing

In [69]:
date_column = sets_node['date'][DATA]
type(date_column)

pandas.core.series.Series

Passing a check for whether the value in the Series [is not null] to _pandas_, as well as specifying the column names we want to get:

[is not null]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.notna.html

In [70]:
sets = data[date_column.notnull()][sets_columns_names]
display(sets.head(15))

fields__oSets__completeSets      fields__oSets__date  \
Erl   0                            0.0  2017-08-02T13:12:09.542   
      70                           1.0  2017-08-09T09:25:11.710   
      140                          1.0  2017-08-09T10:06:25.480   
      210                          2.0  2017-08-09T11:21:01.555   
      350                          2.0  2017-08-09T11:37:20.862   
      490                          1.0  2017-08-22T15:42:05.751   
      560                          2.0  2017-08-17T11:12:02.820   
      700                          2.0  2017-08-17T13:06:30.183   
      840                          1.0  2017-08-22T16:02:50.682   
Tyler 0                            0.0  2017-07-14T12:31:24.194   
      70                           1.0  2017-07-18T10:31:22.263   
      140                          1.0  2017-07-28T13:05:44.673   
      210                          1.0  2017-08-01T15:27:40.174   
      280                          1.0  2017-08-04T10:18:29.654   
      350                          1.0  2017-07-18T10:51:30.195   

          fields__oSets__dateCompare  fields__oSets__desc  \
Erl   0                   2017-08-02                  NaN   
      70                  2017-08-09                  NaN   
      140                 2017-08-09                  NaN   
      210                 2017-08-09                  NaN   
      350                 2017-08-09                  NaN   
      490                 2017-08-22                  NaN   
      560                 2017-08-17                  NaN   
      700                 2017-08-17                  NaN   
      840                 2017-08-22                  NaN   
Tyler 0                   2017-07-14                  NaN   
      70                  2017-07-18                  NaN   
      140                 2017-07-28                  NaN   
      210                 2017-08-01                  NaN   
      280                 2017-08-04                  NaN   
      350                 2017-07-18                  NaN   

           fields__oSets__growthStage fields__oSets__obsName  \
Erl   0                           7.0                  Tyler   
      70                          8.0                  Tyler   
      140                         7.0                  Tyler   
      210                         9.0                  Stean   
      350                         8.0                  Stean   
      490                         8.0                  Mikki   
      560                         8.0              Gabrielle   
      700                         9.0                  Stean   
      840                         8.0                  Mikki   
Tyler 0                           6.0                  Tyler   
      70                          6.0                  Tyler   
      140                         8.0                  Mikki   
      210                         8.0                  Stean   
      280                         8.5                  Stean   
      350                         6.0                  Tyler   

          fields__oSets__results  fields__oSets__totalA1  \
Erl   0                      NaN                     NaN   
      70               RESULTS.5                   164.0   
      140              RESULTS.5                    66.0   
      210              RESULTS.1                     0.0   
      350              RESULTS.1                     5.0   
      490              RESULTS.5                     1.0   
      560              RESULTS.1                   169.0   
      700              RESULTS.1                    78.0   
      840              RESULTS.5                   187.0   
Tyler 0                      NaN                     NaN   
      70               RESULTS.5                     8.0   
      140              RESULTS.5                    37.0   
      210              RESULTS.5                    52.0   
      280              RESULTS.5                    47.0   
      350              RESULTS.5    

Now, we can repeat this technique to get all the other data sections!

### Outputting cleaner data

To finish up, we should refer to that visualization of grouped columns, and start building a repeatable set of steps to grab the data from each section and relate them to each other before continuing on to the next set of nested sections in the DataFrame. As we grab sections and their related sub-sections, we can toss them into another DataFrame that will ultimately represent the same information in a cleaner form.

Because we're doing this programmatically, we can apply those repeatable steps as functions, in a loop. No manual slogging required. Hit play and relax, it'll be done before you can say "ahhh yeee"!

# Analyzing sections and subsections

For each section, I'd like to open with a summary. In my mind, this would include a picture of what's at the node. If any of these exist, let's see:

- Description of this data series
- Names of all related data series grouped here
- Names of all subsections that go deeper

Rather than repeating the same commands for each section, let's make a function that does all that, so we can just call it once for each section.

In [71]:
def summarize_node(node):
    reports = {"Data nodes": child_data(node),
               "Subsections": child_children(node)}
    report_frames = [
        pandas.Series(data=list(value.keys()), name=key)
        for key, value in reports.items()
        if value  # (not empty)
    ]
    if has_data(node):
        display(node[DATA].describe().to_frame())  # looks nicer as frame
    for series in report_frames:
        display(series.to_frame().T)  # easier to read in landscape orientation


# Appendix: Annotated Data Structure

After summarizing, I'll describe the purpose of each section and its relationship to the others.

## Section: `clients`

In [72]:
clients_node = column_dictionary['clients']
summarize_node(clients_node)

0            1      2      3     4
Data nodes  company  displayText  fname  lname  name

## Section: `observers`

In [73]:
observers_node = column_dictionary['observers']
summarize_node(observers_node)

observers
count          3
unique         3
top        Stean
freq           1

## Section: `fields`

In [74]:
fields_node = column_dictionary['fields']
visualize_tree(fields_node, show_leaf_nodes=False)
summarize_node(fields_node)

client/
oSets/
.   oPoints/
.   .   location/
.   .   .   coords/
.   .   observations/
.   .   .   a1/
.   .   .   a2/
.   .   .   a3/
.   .   .   * |/


0     1     2      3     4
Data nodes  crop  date  desc  image  name

0      1
Subsections  client  oSets

### `fields` Subsection: `client`

In [75]:
field_client_node = fields_node['client']
summarize_node(field_client_node)

0            1      2      3     4
Data nodes  company  displayText  fname  lname  name

### `fields` Subsection: `oSets`

In [76]:
osets_node = fields_node['oSets']
visualize_tree(osets_node, show_leaf_nodes=False)
summarize_node(osets_node)

oPoints/
.   location/
.   .   coords/
.   observations/
.   .   a1/
.   .   a2/
.   .   a3/
.   .   * |/


0     1            2     3            4        5   \
Data nodes  completeSets  date  dateCompare  desc  growthStage  obsName   

                 6        7        8        9        10         11  
Data nodes  results  totalA1  totalA2  totalA3  totalA4  totalSets

0
Subsections  oPoints

### `oSets` Subsection: `oPoints`

In [77]:
opoints_node = osets_node['oPoints']
visualize_tree(opoints_node, show_leaf_nodes=False)
summarize_node(opoints_node)

location/
.   coords/
observations/
.   a1/
.   a2/
.   a3/
.   * |/


0     1
Data nodes  id  name

0             1
Subsections  location  observations

#### `oPoints` Subsection: `location`

In [78]:
location_node = opoints_node['location']
visualize_tree(location_node)
summarize_node(location_node)

coords/
.   * accuracy
.   * altitude
.   * altitudeAccuracy
.   * heading
.   * latitude
.   * longitude
.   * speed
* timestamp


0
Data nodes  timestamp

0
Subsections  coords

##### `location` Subsection: `coords`

In [79]:
coords_node = location_node['coords']
summarize_node(coords_node)

0         1                 2        3         4  \
Data nodes  accuracy  altitude  altitudeAccuracy  heading  latitude   

                    5      6  
Data nodes  longitude  speed

#### `oPoints` Subsection: `observations`

In [80]:
observations_node = opoints_node['observations']
visualize_tree(observations_node)
summarize_node(observations_node)

a1/
.   * number
a2/
.   * number
a3/
.   * number
* anum
* complete
* disabled
* eVnum
* enum
* id
* name
* |/
.   * number


0         1         2      3     4   5     6  7
Data nodes  anum  complete  disabled  eVnum  enum  id  name  |

0   1   2  3
Subsections  a1  a2  a3  |

##### `observations` Subsections: `a1`, `a2`, `a3`

In [81]:
summarize_node(observations_node['a1'])

0
Data nodes  number

##### `observations` Subsection: `|` (natural enemies)

In [82]:
natural_enemies_node = observations_node['|']
summarize_node(natural_enemies_node)

fields__oSets__oPoints__observations__|
count                                     3015
unique                                       9
top                                         e7
freq                                       335

0
Data nodes  number

All possible values of `|`, which are the identities of the natural enemies:

In [83]:
pandas.DataFrame(observations_node['|'][DATA].dropna().unique()).T

0   1   2   3   4   5   6   7   8
0  e1  e2  e3  e4  e5  e6  e7  e8  e9